File: v5/ProjectTemplates/VisualBasic/01.SharedProjectTemplate/Lib/PluginCore.vb
Imports System
Imports System.Diagnostics

Namespace $NameSpace$
    Public Enum ImageType
        Pre
        Post
    End Enum

    Public Enum ExecutionModeEnum
        Synchronous = 0
        Asynchronous = 1
    End Enum

    Public Enum ImageTypeEnum
        PreImage = 0
        PostImage = 1
        Both = 2
    End Enum

    Public Enum IsolationModeEnum
        None = 0
        Sandbox = 1
        External = 2
    End Enum

    Public Enum SourceTypeEnum
        Database = 0
        Disk = 1
        Normal = 2
        AzureWebApp = 3
        FileStore = 4
    End Enum

    Public Enum PluginStepOperationEnum
        Activate = 0
        Deactivate = 1
    End Enum

    Public Enum StageEnum
        PreValidation = 10
        PreOperation = 20
        PostOperation = 40
    End Enum

    Public Enum PluginType
        Plugin = 0
        Workflow = 1
        CustomAction = 2
        DataProvider = 3
        CustomApi = 4
    End Enum

    <DebuggerNonUserCode()>
    <AttributeUsage(AttributeTargets.Class, Inherited:=False, AllowMultiple:=True)>
    Public Class CrmPluginRegistrationAttribute
        Inherits Attribute

        Public Sub New(ByVal message As String, ByVal entityLogicalName As String, ByVal stage As StageEnum, ByVal executionMode As ExecutionModeEnum, ByVal filteringAttributes As String, ByVal stepName As String, ByVal executionOrder As Integer, ByVal isolationModel As IsolationModeEnum)
            Me.Message = message
            Me.EntityLogicalName = entityLogicalName
            Me.Stage = stage
            Me.ExecutionMode = executionMode
            Me.FilteringAttributes = filteringAttributes
            Me.Name = stepName
            Me.ExecutionOrder = executionOrder
            Me.IsolationMode = isolationModel
        End Sub

        Public Sub New(ByVal name As String, ByVal friendlyName As String, ByVal description As String, ByVal groupName As String, ByVal isolationModel As IsolationModeEnum)
            Me.Name = name
            Me.FriendlyName = friendlyName
            Me.Description = description
            Me.GroupName = groupName
            Me.IsolationMode = isolationModel
        End Sub

        Public Sub New(ByVal name As String, ByVal message As String, ByVal pluginType As PluginType)
            Me.Name = name
            Me.Message = message
            Me.PluginType = pluginType
        End Sub

        Public Property Id As String = String.Empty
        Public Property Unregister As Boolean = False
        Public Property RunAs As String = String.Empty
        Public Property FriendlyName As String = String.Empty
        Public Property GroupName As String = String.Empty
        Public Property Description As String = String.Empty
        Public Property DeleteAsyncOperation As Boolean = True
        Public Property Offline As Boolean = False
        Public Property Server As Boolean = True
        Public Property Action As PluginStepOperationEnum = PluginStepOperationEnum.Activate
        Public Property IsolationMode As IsolationModeEnum = IsolationModeEnum.Sandbox
        Public Property Message As String = String.Empty
        Public Property EntityLogicalName As String = String.Empty
        Public Property FilteringAttributes As String = String.Empty
        Public Property Name As String = String.Empty
        Public Property ExecutionOrder As Integer = 1
        Public Property Stage As StageEnum = StageEnum.PostOperation
        Public Property ExecutionMode As ExecutionModeEnum = ExecutionModeEnum.Asynchronous
        Public Property UnSecureConfiguration As String = String.Empty
        Public Property SecureConfiguration As String = String.Empty
        Public Property Image1Name As String = String.Empty
        Public Property Image1Alias As String = String.Empty
        Public Property Image1Type As ImageTypeEnum = ImageTypeEnum.PreImage
        Public Property Image1Attributes As String = String.Empty
        Public Property Image2Name As String = String.Empty
        Public Property Image2Alias As String = String.Empty
        Public Property Image2Type As ImageTypeEnum = ImageTypeEnum.PostImage
        Public Property Image2Attributes As String = String.Empty
        Public Property Image3Name As String = String.Empty
        Public Property Image3Alias As String = String.Empty
        Public Property Image3Type As ImageTypeEnum = ImageTypeEnum.PostImage
        Public Property Image3Attributes As String = String.Empty
        Public Property Image4Name As String = String.Empty
        Public Property Image4Alias As String = String.Empty
        Public Property Image4Type As ImageTypeEnum = ImageTypeEnum.PostImage
        Public Property Image4Attributes As String = String.Empty
        Public Property PluginType As PluginType
        Public Property DataSource As String
    End Class

    <DebuggerNonUserCode()>
    <AttributeUsage(AttributeTargets.Assembly, Inherited:=False, AllowMultiple:=False)>
    Public Class DynamicsCrmDevKitPluginManagedIdentityAssemblyAttribute
        Inherits Attribute

        Public Property TenantId As String
        Public Property CertificateFileName As String
        Public Property CertificatePassword As String
        Public Property ApplicationIds As String = String.Empty
    End Class

    <DebuggerNonUserCode()>
    <AttributeUsage(AttributeTargets.Assembly, Inherited:=False, AllowMultiple:=False)>
    Public Class DynamicsCrmDevKitPluginAssemblyAttribute
        Inherits Attribute

        Public Property IsolationMode As IsolationModeEnum = IsolationModeEnum.Sandbox
        Public Property SourceType As SourceTypeEnum = SourceTypeEnum.Database
    End Class
End Namespace

File: v5/DynamicsCrm.DevKit.Tests/TestSdkProjects/Dev.DevKit.Shared/Lib/PluginCore.cs
using System;
using System.Diagnostics;

namespace Dev.DevKit.Shared
{
    internal enum ImageType
    {
        Pre,
        Post
    }

    internal enum ExecutionModeEnum
    {
        Synchronous = 0,
        Asynchronous = 1
    }

    internal enum ImageTypeEnum
    {
        PreImage = 0,
        PostImage = 1,
        Both = 2
    }

    internal enum IsolationModeEnum
    {
        None = 0,
        Sandbox = 1,
        External = 2,
    }

    internal enum SourceTypeEnum
    {
        Database = 0,
        Disk = 1,
        Normal = 2,
        AzureWebApp = 3,
        FileStore = 4
    }

    internal enum PluginStepOperationEnum
    {
        Activate = 0,
        Deactivate = 1
    }

    internal enum StageEnum
    {
        PreValidation = 10,
        PreOperation = 20,
        PostOperation = 40
    }

    internal enum PluginType
    {
        Plugin = 0,
        Workflow = 1,
        CustomAction = 2,
        DataProvider = 3,
        CustomApi = 4
    }

    [DebuggerNonUserCode()]
    [AttributeUsage(AttributeTargets.Class, Inherited = false, AllowMultiple = true)]
    internal class CrmPluginRegistrationAttribute : Attribute
    {
        public CrmPluginRegistrationAttribute(string message, string entityLogicalName, StageEnum stage, ExecutionModeEnum executionMode, string filteringAttributes, string stepName, int executionOrder, IsolationModeEnum isolationModel)
        {
            Message = message;
            EntityLogicalName = entityLogicalName;
            Stage = stage;
            ExecutionMode = executionMode;
            FilteringAttributes = filteringAttributes;
            Name = stepName;
            ExecutionOrder = executionOrder;
            IsolationMode = isolationModel;
        }

        public CrmPluginRegistrationAttribute(string name, string friendlyName, string description, string groupName, IsolationModeEnum isolationModel)
        {
            Name = name;
            FriendlyName = friendlyName;
            Description = description;
            GroupName = groupName;
            IsolationMode = isolationModel;
        }

        public CrmPluginRegistrationAttribute(string name, string message, PluginType pluginType)
        {
            Name = name;
            Message = message;
            PluginType = pluginType;
        }
        public string Id { get; set; } = string.Empty;
        public bool Unregister { get; set; } = false;
        public string RunAs { get; set; } = string.Empty;
        public string FriendlyName { get; set; } = string.Empty;
        public string GroupName { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public bool DeleteAsyncOperation { get; set; } = true;
        public bool Offline { get; set; } = false;
        public bool Server { get; set; } = true;
        public PluginStepOperationEnum Action { get; set; } = PluginStepOperationEnum.Activate;
        public IsolationModeEnum IsolationMode { get; set; } = IsolationModeEnum.Sandbox;
        public string Message { get; set; } = string.Empty;
        public string EntityLogicalName { get; set; } = string.Empty;
        public string FilteringAttributes { get; set; } = string.Empty;
        public string Name { get; set; } = string.Empty;
        public int ExecutionOrder { get; set; } = 1;
        public StageEnum Stage { get; set; } = StageEnum.PostOperation;
        public ExecutionModeEnum ExecutionMode { get; set; } = ExecutionModeEnum.Asynchronous;
        public string UnSecureConfiguration { get; set; } = string.Empty;
        public string SecureConfiguration { get; set; } = string.Empty;
        public string Image1Name { get; set; } = string.Empty;
        public string Image1Alias { get; set; } = string.Empty;
        public ImageTypeEnum Image1Type { get; set; } = ImageTypeEnum.PreImage;
        public string Image1Attributes { get; set; } = string.Empty;
        public string Image2Name { get; set; } = string.Empty;
        public string Image2Alias { get; set; } = string.Empty;
        public ImageTypeEnum Image2Type { get; set; } = ImageTypeEnum.PostImage;
        public string Image2Attributes { get; set; } = string.Empty;
        public string Image3Name { get; set; } = string.Empty;
        public string Image3Alias { get; set; } = string.Empty;
        public ImageTypeEnum Image3Type { get; set; } = ImageTypeEnum.PostImage;
        public string Image3Attributes { get; set; } = string.Empty;
        public string Image4Name { get; set; } = string.Empty;
        public string Image4Alias { get; set; } = string.Empty;
        public ImageTypeEnum Image4Type { get; set; } = ImageTypeEnum.PostImage;
        public string Image4Attributes { get; set; } = string.Empty;
        public PluginType PluginType { get; set; }
        public string DataSource { get; set; }
    }

    [DebuggerNonUserCode()]
    [AttributeUsage(AttributeTargets.Assembly, Inherited = false, AllowMultiple = false)]
    internal class DynamicsCrmDevKitPluginManagedIdentityAssemblyAttribute : Attribute
    {
        public string TenantId { get; set; }
        public string CertificateFileName { get; set; }
        public string CertificatePassword { get; set; }
        public string ApplicationIds { get; set; } = string.Empty;
    }

    [DebuggerNonUserCode()]
    [AttributeUsage(AttributeTargets.Assembly, Inherited = false, AllowMultiple = false)]
    internal class DynamicsCrmDevKitPluginAssemblyAttribute : Attribute
    {
        public IsolationModeEnum IsolationMode { get; set; } = IsolationModeEnum.Sandbox;
        public SourceTypeEnum SourceType { get; set; } = SourceTypeEnum.Database;
    }
}


